Make logging for validation optional

laanwj opened the pull request:
Avoid ERROR messages like `ERROR: AcceptToMemoryPool : non-final`. Logging of incoming transaction validation failures can be enabled optionally with`-debug=mempoolrej`.

Also introduces a consistent format for transaction validation failures that always includes the txid.

```
2015-08-05 13:25:26 txfail 4e5014c4084aab04aac280e4dc6e26ffbc76ea0feb817b98a1a67d8b18d4e80b: AcceptToMemoryPool: inputs already spent
```

I have opted for this approach instead of the one outlined in #5794 to remove all messages, because they provide extra troubleshooting information.

Fixes #5794.

sipa commented:
Hmm, have you considered to not print the error immediately, but only report through ValidationState, and then have ProcessMessage choose (optionally) to report the error string it, if any?

That's also more compatible with future modularization of the consensus code, where you want to do the actual logging externally.

laanwj commented:
Yes let's do that...

laanwj committed:
Add debug message to CValidationState for optional extra information

Add a field `strDebugMessage` which can be passed to DoS or Invalid,
and queried using GetDebugMessage() to add extra troubleshooting
information to the validation state.

laanwj committed:
Introduce REJECT_INTERNAL codes for local AcceptToMempool errors

Add status codes specific to AcceptToMempool procession of transactions.
These can never happen due to block validation, and must never be sent
over the P2P network. Add assertions where appropriate.

laanwj committed:
Add function to convert CValidationState to a human-readable message

It is necessary to be able to concisely log a validation state.
Convert CValidationState to a human-readable message for logging.

laanwj committed:
Remove most logging from transaction validation

Remove unnecessary direct logging in CheckTransaction,
AcceptToMemoryPool, CheckTxInputs, CScriptCheck::operator()

All status information should be returned in the CValidationState.
Relevant debug information is also added to the CValidationState using
the recently introduced debug message.

Do keep the "BUG! PLEASE REPORT THIS! ConnectInputs failed against MANDATORY but not STANDARD flags"
error as it is meant to appear as bug in the log.

laanwj committed:
Add information to errors in ConnectBlock, CheckBlock

Add detailed state information to the errors, as it is no longer being
logged downstream.

Also add the state information to mempool rejection debug message in
ProcessMessages.

laanwj committed:
Move mempool rejections to new debug category

Move mempool rejections to debug category `mempoolrej`, to make it possible
to show them without enabling the entire category `mempool` which is
high volume.

laanwj commented:
Done. CValidationState is augmented with optional debug information, which is logged upstream where possible.

This makes the change larger than originally intended, but the end result is better.
See commit messages/details for more information.

N.B.: I've also removed the logging in `CScriptCheck::operator()()`. This is consistent but I'm uncertain about this, because this means that the script error is no longer logged in case of verification through CCheckQueue (so in the case of blocks). However it *is* logged when called directly from CheckInputs (https://github.com/bitcoin/bitcoin/blob/master/src/main.cpp#L1394).

sipa commented on src/main.cpp:
```diff
@@ -778,6 +769,14 @@ CAmount GetMinRelayFee(const CTransaction& tx, unsigned int nBytes, bool fAllowF
+{
+    return strprintf("%s%s (code %i)",
+        state.GetRejectReason(),
+        state.GetDebugMessage().empty() ? "" : (", "+state.GetDebugMessage()),
```
No closing bracket?

laanwj replied:
The bracket is not in the string.

laanwj replied:
Removed the parenthesis around ", "+... - they made the code harder to read instead of easier.

sipa commented:
@laanwj That's not very consistent, as I think that means it will be printed with -par=1 and not with other values. I don't know of a better solution, though.

laanwj commented:
@sipa So we just call par=1 mode 'safe mode with debugging'. Voila.

sipa commented:
Untested ACK (but see nit about closing bracket).

dexX7 commented:
> Also introduces a consistent format for transaction validation failures that always includes the txid.

Great PR, I'm running it for some hours, and this is really useful!

morcos commented on src/main.h:
```diff
@@ -455,7 +455,14 @@ extern CBlockTreeDB *pblocktree;
+ * be sent over the P2P network.
+ */
+static const unsigned int REJECT_INTERNAL = 0x100;
 /** local "reject" message codes for RPC which can not be triggered by p2p trasactions */
```
This comment isn't accurate is it?

laanwj replied:
Correct. This comment only holds for HIGHFEE.

morcos commented:
I really like this idea, there are going to be a whole slew of new possible reject reasons with 6470 and with limited mempool chains.  It will be nice to be able to output them in debugging situations.

dcousens commented:
concept ACK

skarra commented:
:+1:

droark commented on src/main.h:
```diff
@@ -455,7 +455,16 @@ extern CBlockTreeDB *pblocktree;
+ * for transactions, to signal internal conditions. They cannot and should not
+ * be sent over the P2P network.
+ */
+static const unsigned int REJECT_INTERNAL = 0x100;
```
Just curious, was the duplication of the REJECT_HIGHFEE value intentional?

jtimon commented:
I've been trying to "move error messages up" for consensus and policy functions in different ways and with multiple attempts since at least January 15th 2015 (the first version of #5669 which also moved CheckTransaction to consensus/consensus did that).
I would have appreciated to be asked for review on this one. Now it's too late to complain about its flaws but I still will do it.
